### ISSUE TITLE
Object Rest Spreads not quite working

### ISSUE THREAD

**Issue by robcolburn** [eslint/eslint#3271]
Object Rest Spreads aren't quite right in the esilnt though.  You can enable them with `experimentalObjectRestSpread`, however ESLint still considers the resulting variables as not defined.

```js
function stealTheBacon(stuff) {
  var {bacon, ...others} = stuff;
  console.log('The bacon is mine', bacon);
  return others;
}

stealTheBacon({
  bacon: "tasty",
  lettuce: "nutricious",
  tomatoe: "juicy"
});
```

```json
{
  "env": {
    "es6": true
  },
  "ecmaFeatures": {
    "experimentalObjectRestSpread": true
  },
  "rules": {
    "no-console": 0,
    "no-undef": 2
  },
  "globals": {
    "console": true
  }
}
```

```sh
4:10  error  "others" is not defined  no-undef
```

This is troublesome for React JSX since the feature is recommended by the docs.
https://facebook.github.io/react/docs/transferring-props.html#transferring-with-...-in-jsx

You can somewhat work around this by declaring ahead of time
```js
function stealTheBacon(stuff) {
  var others;
  var {bacon, ...others} = stuff;
  console.log('The bacon is mine', bacon);
  return others;
}
```

You're re-declaring though, so now the linter should be complaining at you.

**Comment by eslintbot** [comment]
Thanks for the issue! We get a lot of issues, so this message is automatically posted to each one to help you check that you've included all of the information we need to help you.

Reporting a bug? Please be sure to include:

1. The version of ESLint you are using (run `eslint -v`)
1. The source code that caused the problem
1. The configuration you're using (for the rule or your entire config file)
1. The actual ESLint output complete with line numbers

Requesting a new rule? Please be sure to include:

1. The use case for the rule - what is it trying to prevent or flag?
1. Whether the rule is trying to prevent an error or is purely stylistic
1. Why you believe this rule is generic enough to be included

Requesting a feature? Please be sure to include:

1. The problem you want to solve (don't mention the solution)
1. Your take on the correct solution to problem

Including this information in your issue helps us to triage it and get you a response as quickly as possible.

Thanks!

**Comment by nzakas** [comment]
I believe this was due to a bug in estraverse, but I can't find the reference now. @mysticatea do you remember this?

**Comment by mysticatea** [comment]
Ah... I guess this needs a patch to make references in escope.
https://github.com/estools/escope/blob/master/src/referencer.js#L50-L65

However, that class is private!

**Comment by nzakas** [comment]
Can you create an issue on escope for that? Let's see if we can figure out a solution.

**Comment by robcolburn** [comment]
@nzakas,
I take it you're referring @mysticatea for that question? I'm afraid I don't understand the boundaries between this project and that one.

**Comment by nzakas** [comment]
Yes :)

**Comment by hzoo** [comment]
Yeah had to make this patch to fix similar issue I think. https://github.com/babel/babel-eslint/blob/master/index.js#L266-L276. I guess a fix in escope/eslint would allow me to remove it!

**Comment by nzakas** [comment]
Filed here: https://github.com/estools/escope/issues/76

**Comment by mysticatea** [comment]
I'm sorry, I had filed escope/issues#74.

But I thought escope is not extensible for ES7 or others. So I guess a correct way to support ExperimentalRest/SpreadProperty is to fork escope.... And for parser plugins, we need a hook of reference analysis?

**Comment by nzakas** [comment]
I'd rather not create a fork, let's see if we can work with @constellation to figure this out. Ideally, we can do a traversal regardless of what is in the AST.

**Comment by gyandeeps** [comment]
@nzakas @mysticatea Please confirm that my label updates are correct?

**Comment by nzakas** [comment]
Yup, correct.

**Comment by mysticatea** [comment]
I sent a PR: estools/escope#78

**Comment by robcolburn** [comment]
I believe this is fixed now :cake: I think `escope` pushed a version to npm w/o pushing tag to GitHub.

**Comment by mysticatea** [comment]
Oh, I had not noticed 3.2.0 since git tag is nothing. Thanks @robcolburn 
I will add some tests to ESLint to close this issue.

**Comment by michaelficarra** [comment]
@mysticatea I just added the `3.2.0` tag for you: https://github.com/estools/escope/tree/3.2.0

**Comment by mysticatea** [comment]
@michaelficarra Thanks! :)

**Comment by mysticatea** [comment]
escope v3.2.0 does not include https://github.com/estools/escope/commit/cebbcf404f34791ea86a7284f99b556233fa5bc3, so this issue is not fixed yet.

**Comment by nzakas** [comment]
@mysticatea I believe v3.2.1 was released, can you check this issue?

**Comment by robcolburn** [comment]
This is no longer an issue for my React projects… but, i've been an alternate config.
```json
  "parser": "babel-eslint",
  "ecmaFeatures": {
    "jsx": true,
  },
  "plugins": [
    "react"
  ],
```

**Comment by mysticatea** [comment]
I checked with ESLint `v1.9.0` and new escope, it's working correctly now!

**Comment by ilyavolodin** [comment]
So this issue can be closed?

**Comment by mysticatea** [comment]
I will add a test for this, so please wait...

**Comment by ilyavolodin** [comment]
Sounds good. Thanks.
